### PR TITLE
[GEN-1729]: add "describe source" to UI

### PR DESCRIPTION
**Pull request by BenElferink**
This pull request introduces a new feature to describe sources in the application, along with several supporting changes across multiple files. The most important changes include adding a new GraphQL query, creating a custom hook to use this query, and updating the UI to display the described source data.

### New Feature: Describe Source

* **GraphQL Query**:
  * Added a new query `DESCRIBE_SOURCE` in `frontend/webapp/graphql/queries/describe.ts` to fetch detailed information about a source.
  * Exported the new query in `frontend/webapp/graphql/queries/index.ts`.

* **Custom Hook**:
  * Created a new hook `useDescribeSource` in `frontend/webapp/hooks/describe/useDescribeSource.ts` to use the `DESCRIBE_SOURCE` query.
  * Exported the new hook in `frontend/webapp/hooks/describe/index.ts` and `frontend/webapp/hooks/index.ts`. [[1]](diffhunk://#diff-4acfd7087b0688dfb3cb2e8adc9cb925048e2976396258273a7459dcdc40c666R1) [[2]](diffhunk://#diff-2a3eb3a8cb54a543dce38afda808b30eff53e0ef144ec46310eddf7e2796f5a7R1-R10)

* **UI Updates**:
  * Updated `frontend/webapp/containers/main/sources/source-drawer-container/index.tsx` to use the new `useDescribeSource` hook and display the described source data in a new `DataCard`. [[1]](diffhunk://#diff-d4745da2f63be90cb1f7af8ea29fb2f37cf0a80f0f04a207b6fc9149855ca6d5L4-R11) [[2]](diffhunk://#diff-d4745da2f63be90cb1f7af8ea29fb2f37cf0a80f0f04a207b6fc9149855ca6d5R96) [[3]](diffhunk://#diff-d4745da2f63be90cb1f7af8ea29fb2f37cf0a80f0f04a207b6fc9149855ca6d5R145-R153)
  * Added a new reusable component `Code` in `frontend/webapp/reuseable-components/code/index.tsx` to render code snippets with syntax highlighting.
  * Updated `frontend/webapp/reuseable-components/data-card/data-card-fields/index.tsx` to support a new field type `CODE` for displaying code snippets. [[1]](diffhunk://#diff-ce146776b402d3e7c41f2ebb82b32a2b8636ec73200d70a6f30f0f08dc1da6d0L3-R11) [[2]](diffhunk://#diff-ce146776b402d3e7c41f2ebb82b32a2b8636ec73200d70a6f30f0f08dc1da6d0R101-R106)
  * Exported the new `Code` component in `frontend/webapp/reuseable-components/index.ts`.

### Utility Functions

* **Flatten Object Keys**:
  * Added a new utility function `flattenObjectKeys` in `frontend/webapp/utils/functions/formatters/flatten-object-keys/index.ts` to flatten nested object keys.
  * Exported the new utility function in `frontend/webapp/utils/functions/formatters/index.ts`.

* **Safe JSON Stringify**:
  * Added a new utility function `safeJsonStringify` in `frontend/webapp/utils/functions/formatters/safe-json-stringify/index.ts` to safely stringify JSON objects.
  * Updated the imports to use the new `safeJsonStringify` function in `frontend/webapp/containers/main/sources/source-drawer-container/index.tsx`.

### Additional Changes

* **Types**:
  * Added new TypeScript interfaces for the `DescribeSource` response in `frontend/webapp/types/describe.ts`.
  * Exported the new types in `frontend/webapp/types/index.ts`.

* **Dependencies**:
  * Added `prism-react-renderer` for syntax highlighting in `frontend/webapp/package.json`.

These changes collectively add the capability to describe sources in the application, enhancing the user experience by providing detailed information about each source.

Screenshot 2024-12-12 at 22 27 23

### CODE REVIEW

File: frontend/webapp/app/layout.tsx [hunk 1]
```diff
@@ -20,14 +20,20 @@ export default function RootLayout({ children }: { children: React.ReactNode })
...
         <meta name='description' content={METADATA.title} />
         <link rel='icon' type='image/svg+xml' href={`/${METADATA.icons}`} />
         <link rel='icon' type='image/x-icon' href='/favicon.ico' />
         <link rel='icon' type='image/png' sizes='16x16' href='/favicon-16x16.png' />
         <link rel='icon' type='image/png' sizes='32x32' href='/favicon-32x32.png' />
         <link rel='apple-touch-icon' sizes='180x180' href='/apple-touch-icon.png' />
         <link rel='manifest' href='/manifest.json' />
+
+        <link
+          rel='stylesheet'
```
Comment (alonkeyval) [inline frontend/webapp/app/layout.tsx:32]: check the font imports in global.css file

File: frontend/webapp/containers/main/sources/source-drawer-container/index.tsx [hunk 2]
```diff
@@ -141,6 +142,15 @@ export const SourceDrawer: React.FC<Props> = () => {
           <ConditionDetails conditions={item.instrumentedApplicationDetails.conditions} />
           <DataCard title={DATA_CARDS.SOURCE_DETAILS} data={cardData} />
           <DataCard title={DATA_CARDS.DETECTED_CONTAINERS} titleBadge={containersData.length} description={DATA_CARDS.DETECTED_CONTAINERS_DESCRIPTION} data={containersData} />
+          <DataCard
+            title={DATA_CARDS.DESCRIBE_SOURCE}
+            data={[
+              {
+                type: DataCardFieldTypes.CODE,
+                value: JSON.stringify({ language: 'json', code: safeJsonStringify(describe) }),
```
Comment (alonkeyval) [inline frontend/webapp/containers/main/sources/source-drawer-container/index.tsx:150]: what about falllback to stringify?

File: frontend/webapp/reuseable-components/code/index.tsx [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+export const Code: React.FC<Props> = ({ language, code, flatten }) => {
+  const str = flatten && language === 'json' ? safeJsonStringify(flattenObjectKeys(safeJsonParse(code, {}))) : code;
+
+  return (
+    <Highlight theme={prismThemes.palenight} language={language} code={str}>
+      {({ getLineProps, getTokenProps, tokens }) => (
+        <pre>
+          {tokens.map((line, i) => (
+            <div key={`line-${i}`} {...getLineProps({ line })}>
+              {/* <span>{i + 1}</span> */}
```
Comment (alonkeyval) [inline frontend/webapp/reuseable-components/code/index.tsx:28]: remove comment

File: frontend/webapp/reuseable-components/code/index.tsx [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+const Token = styled.span`
+  white-space: pre-wrap;
+  overflow-wrap: break-word;
+  opacity: 0.75;
+  font-size: 12px;
+  font-family: ${({ theme }) => theme.font_family.code};
+`;
+
+export const Code: React.FC<Props> = ({ language, code, flatten }) => {
+  const str = flatten && language === 'json' ? safeJsonStringify(flattenObjectKeys(safeJsonParse(code, {}))) : code;
```
Comment (alonkeyval) [inline frontend/webapp/reuseable-components/code/index.tsx:20]: we can move this outside of the component o make it more ruseable

File: frontend/webapp/reuseable-components/data-card/data-card-fields/index.tsx [hunk 5]
```diff
@@ -97,6 +98,12 @@ const renderValue = (type: DataCardRow['type'], value: DataCardRow['value']) =>
       );
     }
 
+    case DataCardFieldTypes.CODE: {
+      const params = safeJsonParse(value, { language: '', code: '' });
```
Comment (alonkeyval) [inline frontend/webapp/reuseable-components/data-card/data-card-fields/index.tsx:102]:
```suggestion
      const params = safeJsonParse(value, { language: '', code: '' });
  const code = params.flatten ? safeJsonStringify(flattenObjectKeys(params.code)) : params.code;

```

File: frontend/webapp/utils/functions/formatters/flatten-object-keys/index.ts [hunk 6]
```diff
@@ -0,0 +1,46 @@
+// const input = {
```
Comment (alonkeyval) [inline frontend/webapp/utils/functions/formatters/flatten-object-keys/index.ts:1]: clean

**Review by alonkeyval — APPROVED** [review]

File: frontend/webapp/reuseable-components/code/index.tsx (hunk 4 above)
Comment (BenElferink) [reply frontend/webapp/reuseable-components/code/index.tsx:20]: This is the place to make it re-usable, placing it outside would then require copy-pasting of this logic throughout other files.

**Pull request merged**